pkg/componentUpdater/componentUpdater.go

```diff
@@ -0,0 +1,41 @@
+package componentUpdater
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type UpdateFunction func(ctx context.Context, c client.Client, scheme *runtime.Scheme) error
+
+func DefaultUpdater(component client.Object, controller metav1.Object) UpdateFunction {
+	return func(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
+		if controller.GetNamespace() == component.GetNamespace() {
+			if err := controllerutil.SetControllerReference(controller, component, scheme); err != nil {
+				return err
+			}
+		}
+
+		if err := c.Patch(ctx, component, client.Apply, client.ForceOwnership, client.FieldOwner(fmt.Sprintf("%s/%s", controller.GetNamespace(), controller.GetName()))); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func DefaultDeleteOrUpdate(component client.Object, controller metav1.Object, deleteCondition bool) UpdateFunction {
```

Comment:
I have an idea (un tested) for better readability 

```go

func For(component client.Object, controller metav1.Object,  conditions ...Conditions) UpdateFunction {
  if len(conditions) == 0 {
     return patcher(component, controller)
  }

  return conditions[0].Updater
}
```

lets us call 
```go
reconciler.For(newFoobar(), ms)
reconciler.For(newFoobar(), ms,  reconciler.DeleteIf{ ms.Spec.Replica >= 1 } )
reconciler.For(newFoobar(), ms,  reconciler.CreateIf{ ms.Spec.Bar >= 1 } )

```

where 
```go
func(d DeleteIf) Updater() UpdateFunction {
   if d.Condition {
     return deleteReconciler
   }
   return patcher
}
```

Reply:
Yeah I think it's something similar as proposed in the https://github.com/rhobs/observability-operator/pull/184#discussion_r931148861 and I like that approach. Not sure about naming though.....

Reply:
While I very much like syntax that is readable and conveys intention, I would probably argue that the go proverb [Clear is better than clever](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=14m35s) takes precedence here.
If we'd write a library for many users, additional cleverness might be warranted, but here we just want to create some resources.

Reply:
For our usecase, I agree it shouldn't matter much at all. It is an internal implementation detail. 

Having said that .. 
I am not sure I would put it under `too clever` category. My take on that talk is that these are good guidelines (not rules) to follow and can be broken. 

I feel the intention is clear and follows a very simple interface - A condition returns the method to reconcile.


Contrast this `reconciler.NewOptionalReconciler(foobar, ms, ms.Spec.Bar >= 1)` over `reconciler.New(foobar, ms, reconciler.DeleterIf{  ms.Spec.Bar >= 1 })` I would argue that the former requires you to understand the implementation detail of the `NewOptionalReconciler` while the latter does not (the implementation detail doesn't matter). 

But like I said, it is an internal implementation detail ...